WalletWasabi/Crypto/ZeroKnowledge/ZkChallenge.cs

```diff
@@ -0,0 +1,30 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class ZkChallenge
+	{
+		public static Scalar Build(GroupElement publicPoint, GroupElement randomPoint)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			Guard.False($"{nameof(randomPoint)}.{nameof(randomPoint.IsInfinity)}", randomPoint.IsInfinity);
+
+			if (publicPoint == randomPoint)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(randomPoint)} should not be equal.");
+			}
+
+			var concatenation = ByteHelpers.Combine(
+				publicPoint.ToBytes(),
+				randomPoint.ToBytes());
+			using var sha256 = System.Security.Cryptography.SHA256.Create();
+			var hash = sha256.ComputeHash(concatenation);
+			var challenge = new Scalar(hash);
```

Comment:
i would say that it cannot be zero if the challenge is zero, given the Prover sends `s = r + e*k` to the Verifier, if `e` is zero then `s` doesn't take `k` (the value that the Prover want to prove knowledge of) and `s` is just a random number.  

------------

I see a problem with these questions, they are driving the code to caring about situations that even the most pessimist cryptographers don't care about because the numbers are absolutely impossible to obtain. The bitcoin network has a huge hash rate and after 10 years of hashing billions a billions of hashes per second no body has found a hash equal to zero. 

I understand that if we hash something and as result we get zero then something is not okay, i got it, but i think this is going too far.